libs/utils/src/http/endpoint.rs

```diff
@@ -148,27 +147,164 @@ impl Drop for RequestCancelled {
 }
 
 async fn prometheus_metrics_handler(_req: Request<Body>) -> Result<Response<Body>, ApiError> {
+    use bytes::{Bytes, BytesMut};
+    use std::io::Write as _;
+    use tokio::sync::mpsc;
+    use tokio_stream::wrappers::ReceiverStream;
+
     SERVE_METRICS_COUNT.inc();
 
-    let mut buffer = vec![];
-    let encoder = TextEncoder::new();
+    // we need to first test this out with vmagent on staging, if it works, we can get turn this to
+    // true or maybe get rid of the fully buffering path altogether
+    const CHUNKED_IS_DEFAULT: bool = false;
+
+    let chunked = super::request::parse_query_param::<_, bool>(&_req, "allow_chunked")?
+        .unwrap_or(CHUNKED_IS_DEFAULT);
+
+    let started_at = std::time::Instant::now();
+
+    if !chunked {
+        let span = info_span!("blocking");
+
+        tokio::task::spawn_blocking(move || {
+            let _span = span.entered();
+            // Currently we take a lot of mutexes while collecting metrics, so it's
+            // better to spawn a blocking task to avoid blocking the event loop.
+            let metrics = metrics::gather();
+            let mut buffer = Vec::new();
+            let encoder = TextEncoder::new();
+            encoder
+                .encode(&metrics, &mut buffer)
+                .map_err(|e| ApiError::InternalServerError(e.into()))?;
+
+            Ok(Response::builder()
+                .status(200)
+                .header(CONTENT_TYPE, encoder.format_type())
+                .body(Body::from(buffer))
+                .unwrap())
+        })
+        .await
+        .map_err(|e| ApiError::InternalServerError(e.into()))
+        .and_then(|e| e)
+    } else {
+        let (tx, rx) = mpsc::channel(1);
+
+        let body = Body::wrap_stream(ReceiverStream::new(rx));
+
+        struct ChannelWriter {
+            buffer: bytes::BytesMut,
+            tx: tokio::sync::mpsc::Sender<std::io::Result<bytes::Bytes>>,
+            written: usize,
+        }
 
-    let metrics = tokio::task::spawn_blocking(move || {
-        // Currently we take a lot of mutexes while collecting metrics, so it's
-        // better to spawn a blocking task to avoid blocking the event loop.
-        metrics::gather()
-    })
-    .await
-    .map_err(|e: JoinError| ApiError::InternalServerError(e.into()))?;
-    encoder.encode(&metrics, &mut buffer).unwrap();
+        impl ChannelWriter {
+            fn new(buf_len: usize, tx: mpsc::Sender<std::io::Result<Bytes>>) -> Self {
+                ChannelWriter {
+                    buffer: BytesMut::with_capacity(buf_len),
+                    tx,
+                    written: 0,
+                }
+            }
+
+            fn flush0(&mut self) -> std::io::Result<usize> {
+                let n = self.buffer.len();
+                if n > 0 {
```

Comment:
This could be an early return, you save an indentation level.

Reply:
Yeah, it grew a bit, can do.

Reply:
In 88a2cda54a98d7a523d9b8759f0bbf91c2f2f51f